docs/CONTRIBUTING.md

```diff
@@ -51,13 +51,42 @@ the mailing list have a tendency to fall through the cracks.
     entries to `src/types.db`, you should talk to us early in the design
     process.
 
+### Labels
+
+PRs need to be categorized into one of three categories:
+
+*   **Feature**
+
+    The code has new behavior. Performance and documentation improvements are
```

Comment:
```suggestion
    The code has new behavior. Also performance and documentation improvements (other than bug fixes) are
```

Reply:
This is not really helpful when trying to differentiate "Feature" and "Fix", because it essentially says "it's a feature, unless it's a fix".

I guess it's probably unlikely that we'll increase the "minor version" just for a documentation change. We can also document doc updates as "fixes" and trivial changes (e.g. fixing a typo) as "unlisted" / "internal" … WDYT?

Reply:
I tried to clarify that:
* Some performance and documentation changes can be labeled as fixes (e.g. to large performance regression, or grossly misleading documentation)
* A given change (within PR) should be tagged only with one label [1]

Would there be some better wording for that?

[1] If PR contains multiple types of changes; features, fixes, and some non-user-visible changes, I assume that PR should include both "Feature" and "Fix" labels, but maybe "Internal" one could be skipped.

=> If PR has multiple labels, including "Fix", patch release should cherry-pick commits from the PR, instead of taking the whole PR?

=> Review should  reject commits that include multiple types, and ask PR to be split if it has commits of multiple types?

(I have one PR with 2 labels, but as Sysman plugin is new, these distinctions aren't really relevant for it as the whole thing is a new "Feature": https://github.com/collectd/collectd/pull/4079)

> I guess it's probably unlikely that we'll increase the "minor version" just for a documentation change. We can also document doc updates as "fixes" and trivial changes (e.g. fixing a typo) as "unlisted" / "internal" … WDYT?

Hm.  If PR fixes documentation _that relates to an earlier code change PR_ (not yet in release), IMHO it should have the same label as the code change, so that it gets picked to the same branch / release type.

Other documentation updates should be safe to include also into patch/fix releases, and typo fixes _are_ fixes...

Note: I'm not sure what's the role of "Internal" changes in deciding whether something goes to a patch/fix release or not.  Will they be included into those or not?

Reply:
> * Some performance and documentation changes can be labeled as fixes (e.g. to large performance regression, or grossly misleading documentation)

I have addressed this by using the word "typically", e.g. *"Performance improvements are **typically** considered features."* My goal is to give guidance to contributors that are unsure. If there's a good reason for labeling a performance improvement as "fix", that is perfectly fine.

> * A given change (within PR) should be tagged only with one label [1]

IMHO the doc already says this:
*"PRs need to be categorized into one of three categories"*

If that is not clear enough, we could add "exactly":
*"PRs need to be categorized into **exactly** one of three categories"*

> => Review should reject commits that include multiple types, and ask PR to be split if it has commits of multiple types?

This. Don't mix concerns; open separate PRs instead. I have added a note to this effect to the doc.

> (I have one PR with 2 labels, but as Sysman plugin is new, these distinctions aren't really relevant for it as the whole thing is a new "Feature": #4079)

The main problem, in my opinion, is that this code hasn't been released yet: is it really a "fix" if nobody is using the code yet? I think the best answer is to get the code into the hands of users. With regard to the label, either way is fine. When we create a release and there is at least one PR that is marked as a feature, the entire release is going to be considered a feature release.

> Hm. If PR fixes documentation _that relates to an earlier code change PR_ (not yet in release), IMHO it should have the same label as the code change, so that it gets picked to the same branch / release type.

As mentioned above, I don't think this matters enough that we should be arguing about it. I'd just use a "docs" prefix in the PR title, which currently maps to "Fix", but your approach also works for me.

> Note: I'm not sure what's the role of "Internal" changes in deciding whether something goes to a patch/fix release or not. Will they be included into those or not?

As mentioned (or possibly implied) in #4222 I'd like to move towards a "trunk based" development model. So the next release will include all new PRs and we don't cherry-pick into release branches.